### ISSUE TITLE
m3gdb

### ISSUE THREAD

**Issue by johnperry-math** [modula3/cm3#41]
What's the status of m3gdb? The docs refer to a script that's supposed to build it, but the script doesn't exist, near as I can tell. I tried compiling it but it didn't work on my machine, and when looking at the error I didn't get why it should ever have compiled. I can provide details if need be, but I thought I'd ask first if it's currently supported.

**Comment by RodneyBates** [comment]
I use it all the time, but only occasionally rebuild it.  My last build is dated May 22, 2018.
This is on AMD64_LINUX.

What docs are you looking at?  cm3/doc/help/m3gdb/m3gdb-onepage.html?  Both methods given in
the second section ("Building (m3)gdb") have been working for me for a long time.

I seem to be the only one doing any maintenance on m3gdb.  I have a long list of needed
improvements and bug fixes, but only work on it as I need, or somebody else requests.
The last change I made was Dec 3, 2016.

There are some targets it isn't building on, though I don't remember which one(s).  The
only ones I have are AMd64_LINUX and LINUXLIBC6.  If somebody can give me command line
access and wants it on another target, I would do it.

**Comment by RodneyBates** [comment]
The path given in cm3/doc/help/m3gdb/m3gdb-onepage for building without using the
M3 build system is wrong.  It should be cm3/m3-sys/m3gdb/gdb.  I just corrected this.

Both methods are failing, trying to build info documentation on bfd.  The build and
source files involved have not changed in years.  It looks like a newer version of
tex or latex undermined it.  Have I ever mentioned that I don't particularly care for
debugging latex code?  Info files for the bfd library are really not needed anyway.

'Til this is fixed, I can send anybody who wants it, a m3gdb executable for AMD64_LINUX
or LINUXLIBC6.

**Comment by mikanystrom** [comment]
Very strange: TeX is archival, it should never change, that's part of the
design.  LaTeX is not quite as good about that, but generally speaking, I
don't have trouble LaTeXing documents that are decades old...is GNU pulling
in some sketchy libraries from somewhere?

**Comment by johnperry-math** [comment]
Let me establish that I'm trying to build this on MacOS with `clang`, so I'm not 100% surprised that it wouldn't compile. (I'm surprised that `cm3` built on 10.13.6; it's not building on another machine with an older OS.) That aside:

> What docs are you looking at? cm3/doc/help/m3gdb/m3gdb-onepage.html? Both methods given in the second section ("Building (m3)gdb") have been working for me for a long time.

The first method says to try `./do-cm3-m3gdb.sh buildship`. When I try that, this is what happens:

    making /Users/user/modula3/cm3/scripts/PKGS (slow but rare)
    /Users/user/modula3/cm3/scripts/pkgmap.sh -c "cm3 -build -override $RARGS -DROOT='/Users/user/modula3/cm3' " m3gdb
    === package /Users/user/modula3/cm3/m3-sys/m3gdb ===
     +++ cm3 -build -override $RARGS -DROOT='/Users/user/modula3/cm3'  +++
    --- building in AMD64_DARWIN ---
    
     ==> /Users/user/modula3/cm3/m3-sys/m3gdb done

Notice that nothing has actually compiled or been built.

The second method chokes, twice. The first time, it chokes on line 102 of `bfd/pef.c`, which has

    if (! table.flags1 & TB_HAS_TBOFF)

`clang` complains that this needs more parentheses, so I tried this, mimicking the similar test in line 99:

    if (! (table.flags1 & TB_HAS_TBOFF))

and then it compiles a while until it gets to `i386-dis.c`, which has a number of constructions of this form:

    i386-dis.c:3042:23: error: adding 'char' to a string does not append to the string [-Werror,-Wstring-plus-int]
          oappend ("%cs:" + intel_syntax);
                   ~~~~~~~^~~~~~~~~~~~~~

In C code you can't add a `char *` to a `static char` (I think that's what `intel_syntax` is). At least, `clang` thinks you can't, anyway, and I've found that `clang` is usually right on these things. :-) I'd be willing to modify the code to see where I can go from there, but `clang`'s hint,

    use array indexing to silence this warning

doesn't make much sense to me. Looking at it, I half wonder if it's suppose to be a two-array string with `intel_syntax` taking the place of the first element of the array?

**Comment by johnperry-math** [comment]
OK, I tried fixing the above (differently than I indicated; I guessed maybe it was supposed to append `intel_syntax` instead) and I encounter something that suggests I need `gcc`.

    ppc-dis.c:88:35: error: arithmetic on a null pointer treated as a cast from integer to pointer is a GNU extension
          [-Werror,-Wnull-pointer-arithmetic]
      info->private_data = (char *) 0 + dialect;
                           ~~~~~~~~~~ ^

I'm willing to keep trying to make this work with `clang`, but does anyone have a suggested fix for this? (I have one in mind: start at `(char *) 1 + ( dialect - 1)` but I'm not familiar with this construction so I don't want to jump to conclusions.)

**Comment by johnperry-math** [comment]
FWIW, switching to gcc8 via MacPorts doesn't work, either. It complains about multiple case statements falling through (fixed by adding a `/* fall through */` comment) and some other things, including now quite a few like this:

    libbfd.h:218:4: error: cast between incompatible function types from 'bfd_boolean (*)(bfd *)' {aka 'int (*)(struct bfd *)'} to 'bfd_boolean (*)(bfd *, asection *)' {aka 'int (*)(struct bfd *, struct bfd_section *)'} [-Werror=cast-function-type]
       ((bfd_boolean (*) (bfd *, asection *)) bfd_true)

**Comment by johnperry-math** [comment]
> Try first doing ./do-cm3-m3gdb.sh realclean.

Thanks! Unfortunately, it doesn't change the result.

**Comment by johnperry-math** [comment]
Tried building it on a Linux computer today and it still won't build. After giving up on the script and trying the second approach, the latest error gives me this:

    ./elf.texi:11: raising the section level of @subsubsection which is too low
    make[3]: *** [Makefile:360: ../.././bfd/doc/bfd.info] Error 1

I tried `./do-cm3-m3gdb.sh realclean` and had the same result.

**Comment by jaykrell** [comment]
I agree just disable building documentation, or make errors building it not fatal.

m3gdb does not likely work on Mac, at least not for a long time (I thought stabs never worked on Mac, but maybe it used to.) It also doesn’t work on Hpux or NT.
Gcc developers want to deprecate stabs but haven’t yet.

**Comment by VictorMiasnikov** [comment]
Some details can be found here:
https://github.com/modula3/cm3/pull/634

**Comment by VictorMiasnikov** [comment]
I have a problem:
```
{{{{
      ../src/runtime/common/RTLinker.m3────────BFD: /usr/lib/x86_64-linux-gnu/libc.so.6:
   │238           INC (mp, ADRSIZE (mp^));                                             │
   │239         END;                                                                   │
   │240                                                                                │
   │241         (* fix the modules' type links *)                                      │
   │242         mp := modules + start * ADRSIZE (RT0.ModulePtr);                       │
   │243         FOR i := start TO stop DO                                              │
   │244           IF (mp^ # NIL) AND (mp^.link_state = LS_Linked) THEN                 │
   │245             TraceModule("FixTypes: type links: ", mp^);                        │
   │246             ResolveTypeLinks (mp^);                                            │
   │247           END;                                                                 │
   │248           INC (mp, ADRSIZE (mp^));                                             │
   │249         END;                                                                   │
   │250                                                                                │
  >│251         RTTypeSRC.FinishObjectTypes ();                                        │
   │252                                                                                │
   │253         (* verify the partial revelations *)                                   │
   │254         mp := modules + start * ADRSIZE (RT0.ModulePtr);                       │
                       BFD: /usr/lib/x86_64-linux-gnu/libc.so.6: invalid relocation type
multi-thre Thread 1398239 In: FixTypes                          251              3dcac4
Invalid binary operation on numbers.
Invalid binary operation on numbers.
Invalid binary operation on numbers.
Invalid binary operation on numbers.
Invalid binary operation on numbers.
Invalid binary operation on numbers.
Invalid binary operation on numbers.
Invalid binary operation on numbers.
(m3gdb)
}}}}

```

>  'Til this is fixed, I can send anybody who wants it, a m3gdb executable for AMD64_LINUX or LINUXLIBC6.

  Can You public worked m3gdb binary in releases on Github?

And public link to source code.

Thanks!

**Comment by VictorMiasnikov** [comment]
= =
```
./elf.texi:11: raising the section level of @subsubsection which is too low
make[3]: *** [Makefile:360: ../.././bfd/doc/bfd.info] Error 1

```
= =

Can be fixed by  https://github.com/modula3/cm3/pull/776

**Comment by VictorMiasnikov** [comment]
I hope, testing of this m3gdb can sense 

https://github.com/VictorMiasnikov/cm3/releases/download/d5.11.9-ZZYYXX-20220208_13-48/cm3-bin-m3gdb-AMD64_LINUX-d5.11.9-AMD64_LINUX.Debian_11.3.0-1_and_.GCC_11.3.0.-2022-02-08_13-48__FIXed__by__VVM__2022-05-18_14-06__AMD64_LINUX__Target__cm3-3be819e__m3gdb.tar.xz

from
https://github.com/VictorMiasnikov/cm3/releases/tag/d5.11.9-ZZYYXX-20220208_13-48
